### PR TITLE
Enable OAuth integration with Forus.jo accounts

### PR DESCRIPTION
**Pull request by saadaltabari**
- Create an OAuth2 authentication backend for Forus.jo
- Seamlessly create accounts for users signing in using Forus.

### CODE REVIEW

**Review by ghost — COMMENTED** [review]
Thanks @saadaltabari. I still need to review the pull request. Before I go into details I have two points:

 - Do we allow Forus.jo users to link their accounts that they registered with Edraak without entering their password? i.e. is it possible to authenticate someone with the email `ahasan@qrf.org` and potentially gain admin access on Edraak without knowing Ali's password?
 - Could you please add some tests to exercise the new functionality? It's likely that the code will break on future releases and tests are pretty helpful in such cases.

**Comment by saadaltabari** [comment]
Hello @OmarEdraak, thank you for taking the time.
- Regarding your first mentioned point; yes, users can link their Forus accounts with Edraak accounts which have the same email address without needing to enter a password. Your concern is absolutely valid. However, when I first worked on the integration I confirmed with Forus that they verify their user's emails. And since we were going for ease of integration with Forus and because we have a contract with them, we decided to rely on Forus in verifying users seeking authentication. If you still believe this poses too much of a threat, please advise. (looping @qutaiba)

- Regarding your second point; for sure, I will implement tests first thing tomorrow.

File: common/djangoapps/edraak_social/forus/pipeline.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+
+@forus_authentication_only
+def associate_account_by_email(backend, details, user, *args, **kwargs):
+    """
+    This pipeline step associates the current social auth with the user which has the
+    same email address in the database.  It defers from social library's associate_by_email
+    implementation, which verifies that only a single database user is associated with the email.
+    """
+    return associate_by_email(backend, details, user, *args, **kwargs)
```
Comment (ghost) [inline common/djangoapps/edraak_social/forus/pipeline.py:22]: I don't think we should allow `is_superuser` or `is_staff` to authenticate via Forus.

File: common/djangoapps/edraak_social/forus/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+
+@forus_authentication_only
+def create_student_profile(user, *args, **kwargs):
+    try:
+        user.profile
+    except UserProfile.DoesNotExist:
+        profile_details = kwargs.get('profile_details', {})
+        profile_details['user_id'] = user.id
+        UserProfile.objects.create(**profile_details)
```
Comment (ghost) [inline common/djangoapps/edraak_social/forus/pipeline.py:67]: I think we could use `UserProfile.objects.get_or_create` here.

File: common/djangoapps/edraak_social/backends/forus.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    def get_redirect_uri(self, state=None):
+        redirect_uri = super(ForUsOAuth2, self).get_redirect_uri(state)
+        next_param = self.data.get('next')
+        if next_param:
+            redirect_uri += ('&' if '?' in redirect_uri else '?') + \
+                   '{0}={1}'.format(
+                       'next',
+                       next_param.replace(' ', '+')
+                   )
```
Comment (ghost) [inline common/djangoapps/edraak_social/backends/forus.py:44]:
```suggestion
            import urllib
            params = {'next' : next_param}
            encoded_params = urllib.urlencode(next_param)
            redirect_uri = '{url}{sep}{params}'.format(
                url=redirect_uri,
                sep=('&' if '?' in redirect_uri else '?'),
                params=encoded_params,
            )
```

File: common/djangoapps/edraak_social/forus/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+from student.models import UserProfile
+
+from .utils import forus_authentication_only
+
+
+@forus_authentication_only
+def save_redirect_param(strategy, *args, **kwargs):
+    strategy.request.session['next'] = \
+        strategy.request.GET.get('next', '').replace(' ', '+')
```
Comment (ghost) [inline common/djangoapps/edraak_social/forus/pipeline.py:12]: We might need to ask Forus to escape the parameter using a function like `next = encodeURIComponent(next)` or `HttpUtility.UrlEncode` in C#

```suggestion
        strategy.request.GET.get('next', '')
```

File: lms/envs/aws.py [hunk 5]
```diff
@@ -1173,3 +1174,7 @@
 ########################## Derive Any Derived Settings  #######################
 
 derive_settings(__name__)
+
+# the hosts that the platform is safe to redirect to
+ALLOWED_REDIRECT_HOSTS = ENV_TOKENS.get(
+        "SAFE_REDIRECT_HOSTS", tuple())
```
Comment (ghost) [inline lms/envs/aws.py]:
```suggestion
ALLOWED_REDIRECT_HOSTS = ENV_TOKENS.get("SAFE_REDIRECT_HOSTS", [])
```

**Review by ghost — APPROVED** [review]
Thanks Saad.

My recommendation for testing to do the following as time allows:

 - Test the forus backend like this: https://github.com/python-social-auth/social-core/blob/master/social_core/tests/backends/test_azuread_b2c.py 
 - Integration tests to test the workflow like you suggested
 - If you like test the `is_superuser`/`is_staff` check

File: common/djangoapps/edraak_social/backends/forus.py (hunk 3 above)
Comment (saadaltabari) [reply common/djangoapps/edraak_social/backends/forus.py:44]: The purpose of replacing ' ' characters with '+' is not to URL encode the next parameter after all the next param will be sent in the body of a POST request. The next parameter is retrieved from the redirect_uri of the original incoming authentication request. The redirect_uri from the original request is URL encoded and for some reason when it is decoded the + signs are mistaken for spaces. This step replaces spaces with + signs once again.

File: common/djangoapps/edraak_social/forus/pipeline.py (hunk 4 above)
Comment (saadaltabari) [reply common/djangoapps/edraak_social/forus/pipeline.py:12]: After reviewing the authentication implementation as well as the steps followed by Forus for generating the next parameter, I can confirm that the next parameters of incoming requests are URI encoded before being sent to Edraak. However, once these the next parameters are received and decoded the '+' character is mistaken for a space character. This step replaces spaces with + signs once again.

**Pull request merged**